README.MD
# TNet 1.6.7

Tiny C++ wrapper for ENet engine.

### How to start

To see the engine in action in Windows, just go to the \bin directories, and start some applications. 
  
To start developing own applications and games with the engine take a look at the 03.Hybrid example in the \source\examples directory.

## How to build from source

### Microsoft Windows 7/10
Using CMake GUI, go into the root directory, indicate the source as the root folder, and the build folder as the target for the CMake output.
The GUI version of CMake was tested with Windows 7 and 10.

### Ubuntu 16.04
Just go into the root directory and run these commands:
```
~$ cmake ./
~$ make
```

## Dependencies

This is a simple wrapper for the ENet Networking engine. Therefore,
to use it you will need the ENet source code, that already is shipped with
this SDK.

To know more about ENet visit:http://enet.bespin.org/
	

## License
```
MIT License

Copyright (c) 2015-2017 Aramis Hornung Moraes

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
SOFTWARE.


TNet may include ENet code
It's license is the following:

Copyright (c) 2002-2015 Lee Salzman

Permission is hereby granted, free of charge, to any person obtaining
a copy of this software and associated documentation files
(the "Software"), to deal in the Software without restriction, including
without limitation the rights to use, copy, modify, merge, publish,
distribute, sublicense, and/or sell copies of the Software, and to permit
persons to whom the Software is furnished to do so, subject to the
following conditions:

The above copyright notice and this permission notice shall be included
in all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY
CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
```

## Contact

If you have problems, questions or suggestions, please visit the 
official homepage of the TNet Engine:
  
https://aramis.bitbucket.io/
  
If you want to contact the author of the engine, please send an IM to
Aramis Hornung Moraes at:
https://plus.google.com/u/1/107642337350816468504


changes.txt
ahmln-x.y.z.a -
	> 

ahmln-1.6.7.0 -
	> CMake scripts were fixed to work on Ubuntu 16.04.
	> Removed Bully Ellection example due to unnecessary complexity.
	> Some minor changes on the readme.txt.


source/TNet/TServer.cpp
//				  07/08/2015
// TinyEnet - Minimalistic Networking engine
// Wrapper for ENet Networking Engine
// Copyright (c) 2015 Aramis Hornung Moraes
//
//

#include "TServer.h"



void create_stack(Stack *S)
{
    S->size = 0;
}
void *stack_top(Stack *S)
{
    if (S->size == 0) {
        fprintf(stderr, "Error: stack empty\n");
        return 0;
    } 

    return S->data[S->size-1];
}
int stack_push(Stack *S, void *d)
{
    if (S->size < TSTACK_MAX)
	{
        S->data[S->size++] = d;
		return 1;
	}
    else
        fprintf(stderr, "Error: stack full\n");
	return 0;
}
void *stack_pop(Stack *S)
{
    if (S->size == 0)
	{
        fprintf(stderr, "Error: stack empty\n");
		return 0;
	}
    else
	{
		S->size--;
		return S->data[S->size];
	//	free(S->data[S->size-1]);
        
	}
}

TNETServer::TNETServer(int port)
{
	// b. Create a host using enet_host_create
	address.host = ENET_HOST_ANY;
	address.port = port;

	server = enet_host_create(&address, 255, 2, 0, 0);

	if (server == NULL) {
		printf("TNETServer> An error occured while trying to create an ENet server host\n\n");
		exit(EXIT_FAILURE);
	}
	// c. Connect and user service
	eventStatus = 1;

	create_stack(&message_stack);

	printf("TNETServer [STATUS]> Server running, port: %d\n\n", port);
	
}
TNETServer::~TNETServer()
{
	enet_host_destroy(server);
}

int TNETServer::update(int uSecs)
{
	// c. Connect and user service
	if (1) {
		eventStatus = enet_host_service(server, &enet_event, uSecs);
		// If we had some enet_event that interested us
		while (eventStatus > 0 ) {
			switch(enet_event.type) {
			case ENET_EVENT_TYPE_CONNECT:
				eventStatus = enet_host_service(server, &enet_event, uSecs);
					//printf("(Server) We got a new connection from %x\n", enet_event.peer->address.host);
				if(enet_event.type == ENET_EVENT_TYPE_RECEIVE)
					return 1;
				else
					return 2;
					break;

				case ENET_EVENT_TYPE_RECEIVE:
					//printf("(Server) Message from client : %s\n", enet_event.packet->data);
					// Lets broadcast this message to all
					//enet_host_broadcast(server, 0, enet_event.packet);
					return 1;
					break;

				case ENET_EVENT_TYPE_DISCONNECT:
					//printf("%s disconnected.\n", enet_event.peer->data);
					// Reset client's information
					enet_event.peer->data=NULL;
					return 3;
					//break;
			}
		}
	}
	return 0;
}

// WARWNING! DO NOT modify this variable directly, dont forget to
// make a copy of this variable, this is a direct reference to the
// message data packat.
char* TNETServer::getMessage()
{
	return (char*)enet_event.packet->data;
}

TPacket TNETServer::getPacket()
{
	TPacket packet;
	t_packet pckt;
	pckt.event = enet_event;
	packet.setPacket(pckt);
	return packet;
}

void TNETServer::setPacket(TPacket pckt)
{
	// it may be confusing, but we are interested in the event here,
	// since enet suport event. Remember, this is a wrapper, it is
	// supposed  to support other engines, that dont work with Events
	// but packets.
	enet_event = pckt.getPacket().event;
}

int TNETServer::stackPushMessage(char *msg)
{
	return stack_push(&message_stack, msg);
}

void* TNETServer::stackPopMessage()
{
	return stack_pop(&message_stack);
}

int TNETServer::stackSize()
{
	return message_stack.size;
}

void *TNETServer::stackHead()
{
	return stack_top(&message_stack);
}


source/examples/03.Hybrid/main.cpp
//				  07/08/2015
// TinyEnet - Minimalistic Networking engine
// Wrapper for ENet Networking Engine
// Copyright (c) 2015 Aramis Hornung Moraes
//
//

#include "TNet.h"
#include <string.h>

// Little example
int main(int argc, char **argv)
{
	int my_port, remote_port;
	char profile;
	char remote_ip[17];
	char my_message[1024];
	remote_ip[16]=0;

	// start the engine
	startTNetEngine();

	// gather info about the network
	printf("Choose profile client[c] or server [s]: "); fflush(stdin); scanf("%c", &profile);

	while(1)
	{
		if(profile == 's')
		{
			printf("My port: "); fflush(stdin); scanf("%d", &my_port);
			TNETServer serv = TNETServer(my_port);
			
			while(1){
				if(serv.update(1)== TEE_MESSAGE)
				{
					printf("[%s]>: %s\n", serv.getPacket().getSenderIP(), serv.getMessage());
				}
			}
		}
		else if(profile == 'c')
		{
			printf("Host ip : "); fflush(stdin); scanf("%s", &remote_ip);
			printf("Host port : "); fflush(stdin); scanf("%d", &remote_port);
			TNETClient clinet = TNETClient(remote_port, remote_ip);
			while(1){
				printf("Sayt> "); fflush(stdin); scanf("%s",&my_message);
				//clinet.send_peer(my_message);
				clinet.send(my_message, remote_port, remote_ip, strlen(my_message));
			}
			break;
		}
		else
		{
			printf("EXIT\n");
			break;
		}
	}
	return 0;
}

source/TNet/TClient.cpp
//				  07/08/2015
// TinyEnet - Minimalistic Networking engine
// Wrapper for ENet Networking Engine
// Copyright (c) 2015 Aramis Hornung Moraes
//
//

#include "TClient.h"
#include <string.h>
TNETClient::TNETClient(int port, char * server_ip) {
	// b. Create a host using enet_host_create
	client = 0;
	client = enet_host_create(NULL, 1, 2, 57600/8, 14400/8);

	if (client == NULL) {
		printf("TNETClient> An error occured while trying to create an ENet server host\n\n");
	}
	/*
	enet_address_set_host(&address, server_ip);
	address.port = port;

	// c. Connect and user service
	peer = enet_host_connect(client, &address, 2, 0);

	if (peer == NULL) {
		printf("TNETClient> No available peers for initializing an ENet connection\n\n");
		exit(EXIT_FAILURE);
	}
	*/
}

TNETClient::~TNETClient()
{
	enet_host_destroy(client);
}

int TNETClient::update(int uSecs)
{
	if (enet_host_service(client, &event, uSecs) > 0) {
		switch(event.type) {
			case ENET_EVENT_TYPE_RECEIVE:
				//printf("(Client) Message from server : %s\n", event.packet->data);
				// Lets broadcast this message to all
				// enet_host_broadcast(client, 0, event.packet);
				enet_packet_destroy(event.packet);
				return 1;
				break;

			case ENET_EVENT_TYPE_CONNECT:
				//printf("(Client) We got a new connection from %x\n",event.peer->address.host);
				return 2;
				break;

			case ENET_EVENT_TYPE_DISCONNECT:
				//printf("(Client) %s disconnected.\n", event.peer->data);
				// Reset client's information
				event.peer->data = NULL;
				return 3;
				break;
		}
	}
}

int TNETClient::send_peer(char* message, int len)
{
	//printf("Say > ");
	//gets(message);
	if(!len)
		len = strlen(message);
	if (strlen(message) > 0 || len > 0) {
		ENetPacket *packet = enet_packet_create(message, len + 1, ENET_PACKET_FLAG_RELIABLE);
		return enet_peer_send(peer, 0, packet);
	}
	return 0;
}

int TNETClient::send(char* message, int port, char *ip, int len)
{
	ENetAddress taddress;
   // ENetHost *tclient;
    //ENetPeer *tpeer;
    ENetEvent tevent;
	enet_host_service (client, & event, 10); // process any previous message.
    /*tclient = enet_host_create(NULL, 1, 2, 57600/8, 14400/8);

    if (tclient == NULL) {
        fprintf(stderr, "An error occured while trying to create an ENet server host\n");
		enet_host_destroy(tclient);
        return 0;
    }*/
    enet_address_set_host(&taddress, ip);
    taddress.port = port;

    // c. Connect and user service
    peer = enet_host_connect(client, &taddress, 2, 0);
    if (peer == NULL) {
        printf("No available peers for initializing an ENet connection");
		//enet_peer_reset(peer);
        //enet_host_destroy(client);
        return 0;
    }
    //enet_host_service(client, &tevent, 100);
	
	// If we had some event that interested us
	if (enet_host_service(client, &tevent, 128) > 0) {
		if(tevent.type == ENET_EVENT_TYPE_CONNECT)
		{
			if (len > 0) {
				ENetPacket *packet = enet_packet_create(message, strlen(message) + 1, ENET_PACKET_FLAG_RELIABLE);
				//printf("seding message...\n");
				enet_peer_send(peer, 0, packet);
				enet_host_service(client, &tevent, 0);
				//enet_peer_reset(peer);

				//enet_host_destroy(client);


				// DISCONNECT THE DAMN PEER U GOOF
				printf("Sending disconnect packet.\n");
				enet_peer_disconnect (peer, 0);
				/* Allow up to 3 seconds for the disconnect to succeed
				 * and drop any packets received packets.
				 */
				enet_host_service (client, & tevent, 10);
				enet_peer_reset (peer);
				printf("Message sendm peed disconnected..\n");
				return 1;
					/*
				while (enet_host_service (client, & event, 10) > 0)
				{
					switch (event.type)
					{
					case ENET_EVENT_TYPE_RECEIVE:
						enet_packet_destroy (event.packet);
						break;
					case ENET_EVENT_TYPE_DISCONNECT:
						puts ("Disconnection succeeded.\n"); // Debug message TODO: hide later
						return 1;
					}
				}
				/* We've arrived here, so the disconnect attempt didn't */
				/* succeed yet.  Force the connection down.             */
				
				//enet_peer_reset (peer);
			}	
			
		}
	}
	else
	{
		printf("send [error]> Couldn't communicate with destination.\n");
		enet_host_service(client, &tevent, 0);
	}
	enet_peer_reset(peer);
	return 0;
}
